### ISSUE TITLE
Parametrize

### ISSUE THREAD

**Issue by faresobeid** [lucidrains/nGPT-pytorch#1]
Can u explain why parametrize is being used here? Why not just normalise the weights at every step normally without it?

**Comment by lucidrains** [comment]
extra insurance, in the case someone forgets to do that

**Comment by lucidrains** [comment]
I'll make it configurable

**Comment by lucidrains** [comment]
@faresobeid hey, you should be able to turn it off and manually called `model.norm_weights_()` now [as so](https://github.com/lucidrains/nGPT-pytorch/commit/9779df053d1872123ec509d52d528a53e711859a#diff-ed183d67207df065a11e1289f19d34cc2abbc5448dea952683cfe9728c342b95R94)

**Comment by faresobeid** [comment]
Nice thx, im expecting it doesnt affect speed or efficiency right? Also on a different note, if ur using tied embeddings is it recommended to define them seperately then make the weights equal or just define one and use its weight for both embed and unembed, and does it impact the nGPT assumptions or mess up the code?

**Comment by lucidrains** [comment]
they didn't use tied embeddings, and I've never gotten that great results using tied embeddings anyways

It isn't used that much in practice even if it makes a ton of sense

**Comment by lucidrains** [comment]
@faresobeid i can throw in the feature if you let me know your results? are you running some battery of experiments for nGPT?

**Comment by faresobeid** [comment]
I was planning to test it on https://github.com/KellerJordan/modded-nanogpt/tree/master which already uses tied embeddings so didnt want to deviate away from it too much. But its fine, i think it can be implemented easily was just wondering why nanogpt and many other ppl define an lm head and emb then tie them instead of defining one of them and if its an efficiency thing. Ill let u know my results either way

**Comment by lucidrains** [comment]
@faresobeid ohh, you got me curious and so i decided to just try it and it seems to run better than unnormalized gpt

ok, i'm adding it!

**Comment by lucidrains** [comment]
@faresobeid that's the last of my energy today, have fun with training

**Comment by faresobeid** [comment]
Thanks :). Btw to clarify did u mean trying tied embeddings or nGPT in general?

**Comment by lucidrains** [comment]
@faresobeid both actually

usually don't see tied embeddings converging this nicely with unnormalized gpt. loss starts off a bit high, but nothing a bit of better init cannot fix

**Comment by faresobeid** [comment]
Thanks a lot! If u have a bit of energy left (or tmrw lol), what do u think nGPT would look like if using a normal gelu or relu MLP compared to swiglu? Im guessing u just dont have that sqrt(d) factor but idk maybe theres a bit more to it

**Comment by faresobeid** [comment]
Also i dont see this part of the paper implemented 
<img width="815" alt="Screenshot 2024-10-09 at 23 17 15" src="https://github.com/user-attachments/assets/fd345175-3a53-4dee-ae37-30a46193bcf7">

**Comment by lucidrains** [comment]
@faresobeid i'm tired, but also a perfectionist.. ok, i think i fixed the scaling in 0.0.11, can you double check all the numbers for me?

as for using another activation function, i think you would need to apply the same analysis in Appendix 1 for some different constant

**Comment by lucidrains** [comment]
feel like all the messing around with effective learning rates is kind of the lowest yield of the paper, but i'll make the init and scale customizable later

if there's anything to the paper, it would be the strict constraint to the hypersphere

**Comment by faresobeid** [comment]
Ok makes sense. For ur constants it seems like here the init is a bit different (0.05)
<img width="757" alt="Screenshot 2024-10-10 at 00 34 07" src="https://github.com/user-attachments/assets/3ccdce7e-736c-46f6-939a-c373a732d80d">

**Comment by faresobeid** [comment]
> if there's anything to the paper, it would be the strict constraint to the hypersphere
Ya i agree although tbh i ran my first test and was not impressed lol but that was with all scalar fixed values and i mightve done something wrong. Im also tired so tmrw will try to get this working hopefully

**Comment by lucidrains** [comment]
@faresobeid yea, i think they chose that 0.05 because it is around 1 / 24 layers, so i put that logic [here](https://github.com/lucidrains/nGPT-pytorch/blob/main/nGPT_pytorch/nGPT.py#L217), iiuc

**Comment by faresobeid** [comment]
Oh didn’t see that, ya makes sense

**Comment by lucidrains** [comment]
yea, if it actually converges faster like they said, that would be great

i mostly care if it can be scaled up without any noticeable deficiencies

if it could, then it should remove the instability that occurs before the final norm in usual pre-norm transformers

**Comment by lucidrains** [comment]
seems to be running nicely on my end! ok be back tomorrow to cleanup any remaining issues

**Comment by faresobeid** [comment]
Ok im back to try this, could u maybe send some of ur graphs or results?

**Comment by faresobeid** [comment]
Also im wondering if that optimiser in modded nanogpt is messing with it so will revert to adamW fully for now

**Comment by faresobeid** [comment]
Ok running full nGPT (just no qk norm) but with a different optimizer (adamW on emb and all vector or scalars, other optimizer on matrix weights)

**Comment by faresobeid** [comment]
Also didnt change lr or anything like that (no weight decay warmup and grad clip tho)

**Comment by faresobeid** [comment]
Not going great, its like 0.1 loss higher than normal consistently, wheres my 4-20x :(

**Comment by lucidrains** [comment]
@faresobeid haha, i was never expecting anything as high as 20x. i'm just surprised it converges at all, with this amount of normalization. that alone may still make it worth investigating further

what is your learning rate? i found that nGPT can tolerate up to 10x (but still diverges at around `1e-2`)

also, you don't need weight decay on the embedding matrix (or anything for that matter)? that too is normalized

**Comment by faresobeid** [comment]
learning rate is 0.0036 for embedding and 0.00036 for the rest (this is with another optimizer so didnt tune further). Will mess around with this another time but one interesting thing was that loss very quickly started dropping in the linear warmdown stage so maybe that suggests some lr scheduler or velue tweaking is needed

**Comment by lucidrains** [comment]
@faresobeid sounds good, keep me updated

**Comment by lucidrains** [comment]
@faresobeid although the network architectures differ a bit, Karras had a section regarding optimizer / schedulers in his EDM2 (which is also heavily normalized) paper

think he proposed using the [inv sqrt decay scheduler](https://github.com/lucidrains/denoising-diffusion-pytorch/blob/main/denoising_diffusion_pytorch/karras_unet.py#L691). wonder if something similar applies here, i'll have to revisit that section of the paper

**Comment by faresobeid** [comment]
Oh ok thx
